bulk_action.go

```diff
@@ -0,0 +1,106 @@
+package opensearchtools
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Action wraps individual actions on a BulkRequest.
+type Action interface {
+	// GetAction returns a json byte array for the action and metadata of the operation.
+	GetAction() ([]byte, error)
+
+	// GetDoc returns the optional byte array for the document in the operation.
+	// returns nil if no document is needed
+	GetDoc() ([]byte, error)
+}
+
+// ActionResponse encapsulates the individual response for each operation
+type ActionResponse struct {
+	Type        string       `json:"-"`
+	Index       string       `json:"_index"`
+	ID          string       `json:"_id"`
+	Version     uint64       `json:"_version"`
+	Result      string       `json:"result"`
+	Shards      *ShardMeta   `json:"_shards,omitempty"`
+	SeqNo       uint64       `json:"_seq_no"`
+	PrimaryTerm uint64       `json:"_primary_term"`
+	Status      int          `json:"status"`
+	Error       *ActionError `json:"error"`
+}
+
+// UnmarshalJSON implements [json.Unmarshaler] to decode a json byte slice into an ActionResponse
+func (o *ActionResponse) UnmarshalJSON(m []byte) error {
+	// map[action type] -> map[response attribute] -> value
+	var rawResp map[string]map[string]json.RawMessage
+	if err := json.Unmarshal(m, &rawResp); err != nil {
+		return err
+	}
+
+	if len(rawResp) > 1 {
+		return fmt.Errorf("unexpected number of operation responses %d expected 1", len(rawResp))
+	}
+
+	for opType, attrMap := range rawResp {
+		o.Type = opType
+
+		if err := readField(&o.Index, "_index", attrMap); err != nil {
+			return err
+		}
+
+		if err := readField(&o.ID, "_id", attrMap); err != nil {
+			return err
+		}
+
+		if err := readField(&o.Version, "_version", attrMap); err != nil {
+			return err
+		}
+
+		if err := readField(&o.Result, "result", attrMap); err != nil {
+			return err
+		}
+
+		if err := readField(&o.Shards, "_shards", attrMap); err != nil {
+			return err
+		}
+
+		if err := readField(&o.SeqNo, "_seq_no", attrMap); err != nil {
```

Comment:
This is passing a `**ShardMeta` to `readField()`.   Is that intentional?

Reply:
Yes, kinda. I need the `**` if I don't instantiate `ShardMeta` before hand. So the two options are 
```
if err := readField(o.Shards, "_shards", attrMap); err != nil {
	return err
}
```
or
```
if _, contains := attrMap["_shards"]; contains {
  o.Shards = &ShardMeta{}
  if err := readField(o.Shards, "_shards", attrMap); err != nil {
    return err
  }
}
```

Do you have a preference?

Reply:
In hindsight I was being overly pedantic.  `json.Unmarshal()` supports decoding into a `**SomeType`.